contrib/DC-GAN.ipynb

```diff
@@ -0,0 +1,525 @@
+    "!python pytorch-xla-env-setup.py --version $VERSION"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "# Generative Adersarial Networks (GANs)\n",
+    "\n",
+    "In the landmark paper [Goodfellow et al.](https://arxiv.org/abs/1406.2661) , published in 2014, authors introduced this novel paradigm for generative models. The fundamental idea proposed in the work is to train a Generator Network in adversarial setup, where a discriminator network downstream critiques the generated samples.\n",
+    "\n",
+    "Simply put, generator network generates a sample and discriminator network classifies it as a real or fake. Discriminator is also provided with real samples. The objective functions takes the following form:\n",
+    "\n",
+    "$$\\underset{G}{\\text{minimize}}\\; \\underset{D}{\\text{maximize}}\\; \\mathbb{E}_{x \\sim p_\\text{data}}\\left[\\log D(x)\\right] + \\mathbb{E}_{z \\sim p(z)}\\left[\\log \\left(1-D(G(z))\\right)\\right]$$\n",
+    "where <br>\n",
+    "$x \\sim p_\\text{data}$ are samples from the input data. <br>\n",
+    "$z \\sim p(z)$ are the random noise samples.<br>\n",
+    "$G(z)$ are the generated images using the neural network generator $G$, and $D$ is the output of the discriminator, specifying the probability of an input being real."
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Training Setup\n",
+    "\n",
+    "In this example illustrates distributed (data parallel) training of DC-GAN model using MNIST dataset on a TPU device. A TPU device consistes of 4 chips (8 cores; 2 cores/chip). Both the discriminator and generator replica are created on each of 8 cores. The dataset is splitted across the 8 cores. \n",
+    "\n",
+    "<br> At every training step, each of the cores perfoms the forward (loss computation) and backward (gradient computation) on the given minibatch and then [all_reduce](https://www.tensorflow.org/xla/operation_semantics#allreduce) is performed across TPU cores to update the parameters. Notice `xm.optimizer_step` call in the discriminator and optimizer train steps.\n",
+    "\n",
+    "\n",
+    "General GAN training looks like:\n",
+    "\n",
+    "1. update the **generator** ($G$) to minimize the probability of the __discriminator making the correct choice__. \n",
+    "2. update the **discriminator** ($D$) to maximize the probability of the __discriminator making the correct choice__.\n",
+    "\n",
+    "We will use a different objective when we update the generator: maximize the probability of the **discriminator making the incorrect choice**. This small change helps to allevaiate problems with the generator gradient vanishing when the discriminator is confident. This is the standard update used in most GAN papers, and was used in the original paper from [Goodfellow et al.](https://arxiv.org/abs/1406.2661). \n",
```

Comment:
typo: alleviate

Reply:
also, what is the change here? minimize p(correct) is equivalent to maximize p(incorrect) = 1-p(correct), no?

Reply:
It is little different. Log probability of the discriminator making correct choice means E[log(D(x))] E[1- log(D(G(z))], ie classifying real as real and call out fake as fake.
In the discriminator objective, we are maximizing this term.
In the generator objective we can either minimizer the same term or just maximize E[log(D(G(z))], the difference is that in generator step, we dont care about what discriminator is doing on the real images. We are just concerned with maximizing the the probability of fooling the Discriminator in classifying fake as real. That's why in BCE loss, we pass all 1's as the target corresponding to discriminator logits coming from fake data.

The difference is also that when we pass the fake data in the second objective function we detach it, but for the first objective function we dont.

Reply:
ok I read a little bit about it. It's mathematically equivalent to maximize p(incorrext) and minimize p(correct), but in one case, the gradients for G are small and vanishing, and in the other, they are larger, allowing for G to train well initially.